src/App.tsx

```diff
@@ -1,61 +1,133 @@
+import React, { useState } from 'react';
 import './App.scss';
+import { User } from './types/User';
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+import { Todo } from './types/Todo';
+import { TodoList } from './components/TodoList';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+function getUser(userId: number): User | null {
+  const foundUser = usersFromServer.find(user => user.id === userId);
+
+  return foundUser || null;
+}
+
+function getTodosFromServer(): Todo[] {
+  return todosFromServer.map(todo => ({
+    ...todo,
+    user: getUser(todo.userId),
+  }));
+}
+
+export const App: React.FC = () => {
+  const [todos, setTodos] = useState(getTodosFromServer());
+  const [title, setTitle] = useState('');
+  const [userId, setUserId] = useState(0);
+  const [titleError, setTitleError] = useState(false);
+  const [userError, setUserError] = useState(false);
+
+  function getNewTodoId(): number {
+    return (
+      1 +
+      todos.reduce((maxId, todo) => {
+        const { id } = todo;
+
+        return id > maxId ? id : maxId;
+      }, 0)
+    );
```

Comment:
why you are using here reducer?
You just need to find maxId, you have better methods for this

Reply:
I'm using reducer because it's the fastest method I know. It has time complexity of O(n). Getting first element of sorted array has time complexity of O(nlog(n)). For Math.max with map method inside it is: O(2n). The other problem with Math.max is that the 'Maximum call stack size exceeded' exception can be thrown if the array is too big because all array elements are passed as parameters. What else could I use to not sacrifice efficiency? Or maybe in this particular case efficiency isn't that important so we can focus more on code style?